grants_tagger/evaluate_pubs.py

```diff
@@ -0,0 +1,127 @@
+import json
+import typer
+import torch
+
+from transformers import AutoModel, AutoTokenizer
+from tqdm import tqdm
+from sklearn.metrics import classification_report
+from sklearn.preprocessing import MultiLabelBinarizer
+from rich import print
+
+evaluate_pubs_app = typer.Typer()
+
+
+@torch.no_grad()
+def evaluate_pubs(
+    data_path: str, threshold: float = 0.5, batch_size: int = 4, device: str = "cpu"
+):
+    if not torch.cuda.is_available() and device == "cuda":
+        print("[bold red]CUDA is not available, using CPU")
+        device = "cpu"
+
+    with open(data_path, "r") as f:
+        data = json.load(f)
+
+    # Filter publications without tags
+    data = [pub for pub in data if pub["mesh_terms"]]
+
+    tokenizer = AutoTokenizer.from_pretrained("Wellcome/WellcomeBertMesh")
+    model = AutoModel.from_pretrained(
+        "Wellcome/WellcomeBertMesh", trust_remote_code=True
+    )
+    model.to(device)
+    model.eval()
+
+    id2label = model.id2label
+    label2id = {v: k for k, v in id2label.items()}
+    mlb = MultiLabelBinarizer()
+    mlb.fit([list(label2id.values())])
+
+    # Dump labels to a file
+    with open("labels.txt", "w") as f:
+        for label in label2id:
+            f.write(label + "\n")
+
+    # Generate Y_true
+    Y_true = []
+
+    num_ignored_tags = 0
+
+    for pub in data:
+        id_list = []
+        for label in pub["mesh_terms"]:
+            if "/" in label:
+                label = label.split("/", 1)[0]
+
+            if label.endswith("*"):
+                label = label[:-1]
+
+            if label not in label2id:
+                num_ignored_tags += 1
+                continue
+
+            id_list.append(label2id[label])
+        Y_true.append(id_list)
+
+    # Generate Y_pred
+    pbar = tqdm(total=len(data))
+    Y_pred = []
+
+    for idx in range(0, len(data), batch_size):
+        batch = data[idx : idx + batch_size]
+        batch_abstracts = [pub["abstract"] for pub in batch]
+        inputs = tokenizer(
+            batch_abstracts,
+            padding="max_length",
+            truncation=True,
+            max_length=512,
+            return_tensors="pt",
+        )
+
+        inputs = inputs.to(device)
+
+        outs = model(**inputs, return_labels=False)
```

Comment:
you are also not using the predict function so this is yet another duplication, we dont want to rewrite predict code for this mode multiple times.